File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/service/DatasetProcessingService.java
/**
 * Shanoir NG - Import, manage and share neuroimaging data
 * Copyright (C) 2009-2019 Inria - https://www.inria.fr/
 * Contact us on https://project.inria.fr/shanoir/
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/gpl-3.0.html
 */

package org.shanoir.ng.processing.service;

import org.apache.solr.client.solrj.SolrServerException;
import org.shanoir.ng.processing.model.DatasetProcessing;
import org.shanoir.ng.shared.exception.EntityNotFoundException;
import org.shanoir.ng.shared.exception.MicroServiceCommunicationException;
import org.shanoir.ng.shared.exception.RestServiceException;
import org.shanoir.ng.shared.exception.ShanoirException;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.validation.BindingResult;

import java.io.IOException;
import java.util.List;
import java.util.Optional;

/**
 * DatasetProcessing service.
 *
 * @author amasson
 *
 */
public interface DatasetProcessingService {
	
    /**
     * Save an entity.
     *
     * @param entity the entity to create.
     * @return created entity.
     */
    @PreAuthorize("hasAnyRole('ADMIN', 'EXPERT', 'USER') and #entity.getId() == null")
    DatasetProcessing create(DatasetProcessing entity);

    /**
     * Update an entity.
     *
     * @param entity the entity to update.
     * @return updated entity.
     * @throws EntityNotFoundException
     * @throws MicroServiceCommunicationException 
     */
    @PreAuthorize("hasAnyRole('ADMIN', 'EXPERT', 'USER')")
    DatasetProcessing update(DatasetProcessing entity) throws EntityNotFoundException;

    /**
     * Delete an entity.
     *
     * @param id the entity id to be deleted.
     * @throws EntityNotFoundException if the entity cannot be found.
     */
    @PreAuthorize("hasAnyRole('ADMIN', 'EXPERT')")
    void deleteById(Long id) throws ShanoirException, RestServiceException, SolrServerException, IOException;

    /**
     * Unlink given dataset from all processing
     * Remove processing if linked to no other dataset
     *
     * @param datasetId
     */
    @PreAuthorize("hasAnyRole('ADMIN', 'EXPERT')")
    void removeDatasetFromAllProcessingInput(Long datasetId) throws ShanoirException, RestServiceException, SolrServerException, IOException;

    /**
     * Check that the processing object is processable
     *
     * @param processing
     * @throws RestServiceException
     */
    void validateDatasetProcessing(DatasetProcessing processing) throws RestServiceException;

    /**
     * Validate that the DatasetProcessing transformation from DTO to Java Object relative toe the BindingResult is correct
     *
     * @param result
     * @throws RestServiceException
     */
    void validate(BindingResult result) throws RestServiceException;
}

File: shanoir-ng-front/src/app/datasets/shared/dataset-processing.service.ts
/**
 * Shanoir NG - Import, manage and share neuroimaging data
 * Copyright (C) 2009-2019 Inria - https://www.inria.fr/
 * Contact us on https://project.inria.fr/shanoir/
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/gpl-3.0.html
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';

import { EntityService } from '../../shared/components/entity/entity.abstract.service';
import * as AppUtils from '../../utils/app.utils';

import { DatasetProcessing } from './dataset-processing.model';
import { DatasetProcessingDTOService, DatasetProcessingInDTO, DatasetProcessingOutDTO } from './dataset-processing.dto';
import { DatasetDTO, DatasetDTOService } from './dataset.dto';
import { Dataset } from './dataset.model';

@Injectable()
export class DatasetProcessingService extends EntityService<DatasetProcessing> {

    API_URL = AppUtils.BACKEND_API_DATASET_PROCESSING_URL;

    constructor(protected http: HttpClient, 
        private datasetDTOService: DatasetDTOService,
        private datasetProcessingDTOService: DatasetProcessingDTOService) {
        super(http)
        datasetProcessingDTOService.setDatasetProcessingService(this);
    }

	findAllByStudyIdAndSubjectId(studyId: number, subjectId: number): Promise<DatasetProcessing[]> {
		return this.http.get<DatasetProcessingInDTO[]>(this.API_URL + '/study/' + studyId + '/subject/' + subjectId)
            .toPromise().then(dtos => this.mapEntityList(dtos));
	}

    findByInputDatasetId(datasetId: number): Promise<DatasetProcessing[]> {
        return this.http.get<DatasetProcessingInDTO[]>(this.API_URL + '/inputDataset/' + datasetId)
            .toPromise().then(dtos => this.mapEntityList(dtos));
    }

    getInputDatasets(datasetProcessingId: number): Promise<Dataset[]> {
        return this.http.get<DatasetDTO[]>(this.API_URL + '/' + datasetProcessingId + '/inputDatasets/')
            .toPromise().then(dtos => this.datasetDTOService.toEntityList(dtos, [], 'lazy'));
    }

    getOutputDatasets(datasetProcessingId: number): Promise<Dataset[]> {
        return this.http.get<DatasetDTO[]>(this.API_URL + '/' + datasetProcessingId + '/outputDatasets/')
            .toPromise().then(dtos => this.datasetDTOService.toEntityList(dtos, [], 'lazy'));
    }

    get(id: number): Promise<DatasetProcessing> {
        return this.http.get<any>(this.API_URL + '/' + id)
            .toPromise()
            .then(this.mapEntity);
    }

    getEntityInstance() { return new DatasetProcessing(); }

    protected mapEntity = (dto: DatasetProcessingInDTO, quickResult?: DatasetProcessing): Promise<DatasetProcessing> => {
        return this.datasetProcessingDTOService.toEntity(dto, quickResult);
    }

    protected mapEntityList = (dtos: DatasetProcessingInDTO[], result?: DatasetProcessing[]): Promise<DatasetProcessing[]> => {
        if (result == undefined) result = [];
        return this.datasetProcessingDTOService.toEntityList(dtos, result);
    }
    
    public stringify(entity: DatasetProcessing) {
        let dto = new DatasetProcessingOutDTO(entity);
        return JSON.stringify(dto, (key, value) => {
            return this.customReplacer(key, value, dto);
        });
    }
}